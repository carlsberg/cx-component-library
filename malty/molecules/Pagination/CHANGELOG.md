# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- **DSM-210:** removed margin on pagination buttons ([#204](https://github.com/CarlsbergGBS/cx-component-library/issues/204)) ([d1a84e3](https://github.com/CarlsbergGBS/cx-component-library/commit/d1a84e3fd0f8d44aff8e1dbc16ff224196ef5e95))
- **DSM-316:** pagination mobile wrap ([#252](https://github.com/CarlsbergGBS/cx-component-library/issues/252)) ([bf73edf](https://github.com/CarlsbergGBS/cx-component-library/commit/bf73edf1e42630edb6d1d9df6d16b4d1aaae0baf))
- Minor component fixes ([#265](https://github.com/CarlsbergGBS/cx-component-library/issues/265)) ([c75bd7c](https://github.com/CarlsbergGBS/cx-component-library/commit/c75bd7c926163db6efb65ab89a64d8637d7a4664))
- **MS-106:** wrong position of AlertBannar pagination arrow ([#375](https://github.com/CarlsbergGBS/cx-component-library/issues/375)) ([7ce4d58](https://github.com/CarlsbergGBS/cx-component-library/commit/7ce4d58f48409b31c3622c0c9c2ad21bb3705d37))
- **pagination:** [DSM-589] Refactor pagination and add validations ([#629](https://github.com/CarlsbergGBS/cx-component-library/issues/629)) ([caa545d](https://github.com/CarlsbergGBS/cx-component-library/commit/caa545dad76254fcc0893dc4c76cf0792ae40751))
- **pagination:** add background-color white to the input ([#845](https://github.com/CarlsbergGBS/cx-component-library/issues/845)) ([df334d0](https://github.com/CarlsbergGBS/cx-component-library/commit/df334d08b0aa79b1ebb9444940044877dd80a1ad))
- Remove `eslint` warnings and errors ([#322](https://github.com/CarlsbergGBS/cx-component-library/issues/322)) ([4c7bc73](https://github.com/CarlsbergGBS/cx-component-library/commit/4c7bc737de8abb060919ac5ba16220a6d5d0ace9))
- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- **table:** [DSM-667] Update pagination when action on table ([#706](https://github.com/CarlsbergGBS/cx-component-library/issues/706)) ([31a6062](https://github.com/CarlsbergGBS/cx-component-library/commit/31a60623347d0059050ed8c78a77d03f0ecd8c26))
- Update components to have a function declaration and remove unneeded fragments ([#502](https://github.com/CarlsbergGBS/cx-component-library/issues/502)) ([b470095](https://github.com/CarlsbergGBS/cx-component-library/commit/b4700954f4e1ae9c7e547f4de37ac2b15a9d55ea))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-133:** input pagination ([#197](https://github.com/CarlsbergGBS/cx-component-library/issues/197)) ([78595fc](https://github.com/CarlsbergGBS/cx-component-library/commit/78595fc7456b865bb9d1a33dd2003faae323f140))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-229:** button ([#165](https://github.com/CarlsbergGBS/cx-component-library/issues/165)) ([1950717](https://github.com/CarlsbergGBS/cx-component-library/commit/19507172b731f341ecf32c13eac07e465c6aa659))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-285:** table basic version ([#382](https://github.com/CarlsbergGBS/cx-component-library/issues/382)) ([c761462](https://github.com/CarlsbergGBS/cx-component-library/commit/c761462ef7ba7d40c6cb549a050e64ba942b919b))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- New Chip component and minor changes ([#239](https://github.com/CarlsbergGBS/cx-component-library/issues/239)) ([ce92ab0](https://github.com/CarlsbergGBS/cx-component-library/commit/ce92ab0543b96296952a29e2d43d0df29d4e188f))
- **pagination:** [DSM-935] add disabled state ([#836](https://github.com/CarlsbergGBS/cx-component-library/issues/836)) ([6743adc](https://github.com/CarlsbergGBS/cx-component-library/commit/6743adcd8b600d68ea9d03096a66c275dc6e6bb1))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
- Version bumps for various fixes ([#205](https://github.com/CarlsbergGBS/cx-component-library/issues/205)) ([57c2e30](https://github.com/CarlsbergGBS/cx-component-library/commit/57c2e30804ba0aa6ebdd6a3ac59044b69ad9393f))
