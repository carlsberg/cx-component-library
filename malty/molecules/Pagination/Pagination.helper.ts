import { useMemo } from 'react';

export const LEFT_DOTS = 'left-dots';
export const RIGHT_DOTS = 'right-dots';

const range = (start: number, end: number) => {
  const length = end - start + 1;
  return Array.from({ length }, (_, idx) => idx + start);
};

interface Props {
  totalPageCount: number;
  siblingCount?: number;
  currentPage: number;
  isDefault: boolean;
}

export const usePagination = ({ totalPageCount, siblingCount = 1, currentPage, isDefault }: Props) =>
  useMemo(() => {
    if (!isDefault) {
      return null;
    }

    // Pages count is determined as siblingCount + firstPage + lastPage + currentPage + 2*DOTS
    const totalPageNumbers = siblingCount + 5;

    /*
      If the number of pages is less than the page numbers we want to show in our
      paginationComponent, we return the range [1..totalPageCount]
    */
    if (totalPageNumbers >= totalPageCount) {
      return range(1, totalPageCount);
    }

    const leftSiblingIndex = Math.max(currentPage - siblingCount, 1);
    const rightSiblingIndex = Math.min(currentPage + siblingCount, totalPageCount);

    /*
      We do not want to show dots if there is only one position left
      after/before the left/right page count as that would lead to a change if our Pagination
      component size which we do not want
    */
    const shouldShowLeftDots = leftSiblingIndex > 2;
    const shouldShowRightDots = rightSiblingIndex < totalPageCount - 2;

    const firstPageIndex = 1;
    const lastPageIndex = totalPageCount;

    if (!shouldShowLeftDots && shouldShowRightDots) {
      const leftItemCount = 3 + 2 * siblingCount;
      const leftRange = range(1, leftItemCount);

      return [...leftRange, RIGHT_DOTS, totalPageCount];
    }

    if (shouldShowLeftDots && !shouldShowRightDots) {
      const rightItemCount = 3 + 2 * siblingCount;
      const rightRange = range(totalPageCount - rightItemCount + 1, totalPageCount);

      return [firstPageIndex, LEFT_DOTS, ...rightRange];
    }

    if (shouldShowLeftDots && shouldShowRightDots) {
      const middleRange = range(leftSiblingIndex, rightSiblingIndex);

      return [firstPageIndex, LEFT_DOTS, ...middleRange, RIGHT_DOTS, lastPageIndex];
    }

    return null;
  }, [totalPageCount, siblingCount, currentPage, isDefault]);
