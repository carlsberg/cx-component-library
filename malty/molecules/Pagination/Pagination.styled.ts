import styled from 'styled-components';

export const StyledContainer = styled.div<{ isWhite?: boolean }>`
  display: flex;
  justify-content: center;
  align-items: center;
  button {
    svg {
      fill: ${({ theme, isWhite }) =>
        isWhite ? theme.colors.colours.default.white.value : theme.colors.colours.default['digital-black'].value};
    }
    :disabled {
      svg {
        fill: ${({ theme, isWhite }) =>
          isWhite
            ? theme.colors.colours.overlay.white[25].value
            : theme.colors.colours.system['disable-light-theme'].value};
      }
      background-color: transparent;
      :hover {
        background-color: transparent;
      }
    }
  }
  ul {
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    padding: 0;
    margin: 0;
    list-style: none;
    li:first-child {
      & button {
        padding: 0 ${({ theme }) => theme.sizes['2xs'].value};
      }
      margin-right: 8px;
    }
    li:last-child {
      & button {
        padding: 0 ${({ theme }) => theme.sizes['2xs'].value};
      }
      margin-left: ${({ theme }) => theme.sizes['2xs'].value};
    }
  }
`;

export const StyledDots = styled.div<{ isWhite?: boolean }>`
  height: ${({ theme }) => theme.sizes.xl.value};
  width: ${({ theme }) => theme.sizes.xl.value};
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: default;
  color: ${({ theme, isWhite }) =>
    isWhite ? theme.colors['text-colours'].white.value : theme.colors['text-colours']['digital-black'].value};
`;
