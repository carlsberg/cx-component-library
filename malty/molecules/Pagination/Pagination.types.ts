export interface PaginationProps {
  count: number;
  onChange: (page: number) => void;
  currentPage: number;
  siblingCount?: number;
  type?: PaginationType;
}

export enum PaginationType {
  default,
  compact
}
