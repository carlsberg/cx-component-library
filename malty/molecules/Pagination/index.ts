export { Pagination } from './Pagination';
export { PaginationType } from './Pagination.types';
export type { PaginationProps } from './Pagination.types';
