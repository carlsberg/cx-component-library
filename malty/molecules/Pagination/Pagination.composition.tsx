import React, { useState } from 'react';
import { Pagination } from './Pagination';

export const BasicPagination = () => {
  const [page, setPage] = useState(0);
  return <Pagination page={page} onChange={(page) => setPage(page)} />;
};
