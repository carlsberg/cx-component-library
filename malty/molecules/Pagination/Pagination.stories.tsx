import { Meta, StoryObj } from '@storybook/react';
import React, { useState } from 'react';
import { Pagination } from './Pagination';
import { PaginationProps, PaginationType } from './Pagination.types';

const PaginationComponent = ({ ...args }) => {
  const { currentPage, count } = args;
  const [statePage, setStatePage] = useState(currentPage);

  return (
    <Pagination
      {...args}
      count={count}
      onChange={(page) => {
        setStatePage(page as number);
      }}
      currentPage={statePage}
    />
  );
};

const meta: Meta<PaginationProps> = {
  component: PaginationComponent,
  title: 'Navigation/Pagination',
  parameters: {
    importObject: 'Pagination',
    importPath: '@carlsberggroup/malty.molecules.pagination'
  },
  render: (args) => <Pagination {...args} />,
  argTypes: {
    onChange: {
      description: 'Function to be executed when page changes'
    },
    currentPage: {
      description: 'Current number of pagination',
      control: 'number'
    },
    count: {
      description: 'Total number of pages',
      control: 'number'
    },
    siblingCount: {
      description: 'You can specify how many digits pages to display',
      control: 'number'
    },
    dataTestId: {
      control: 'text',
      description: 'Pagination data-testid'
    },
    type: {
      description: 'Type options for pagination',
      options: Object.keys(PaginationType),
      mapping: PaginationType,
      control: {
        type: 'select',
        label: Object.keys(PaginationType)
      }
    },
    isWhite: {
      description: 'Changes color of component to white',
      control: 'boolean'
    },
    disabled: {
      description: 'Property that allows to disable the component',
      control: 'boolean'
    }
  }
};

type Story = StoryObj<PaginationProps>;

export const Base: Story = {
  args: {
    count: 10,
    currentPage: 1,
    type: PaginationType.Default,
    disabled: false
  }
};

export const Compact: Story = {
  args: {
    ...Base.args,
    type: PaginationType.Compact
  }
};

export const Input: Story = {
  args: {
    ...Base.args,
    type: PaginationType.Input
  }
};

export const Disabled: Story = {
  args: {
    ...Base.args,
    disabled: true
  }
};

export default meta;
