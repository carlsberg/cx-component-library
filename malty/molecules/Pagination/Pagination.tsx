import { Button, ButtonSize, ButtonStyle } from '@carlsberggroup/malty.atoms.button';
import { IconName } from '@carlsberggroup/malty.atoms.icon';
import { Text, TextColor, TextStyle } from '@carlsberggroup/malty.atoms.text';
import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import { EventKeys } from '@carlsberggroup/malty.utils.consts';
import React, { ChangeEvent, FocusEvent, KeyboardEvent, useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { LEFT_DOTS, RIGHT_DOTS, usePagination } from './Pagination.helper';
import { StyledContainer, StyledDots, StyledInput, StyledInputPagination } from './Pagination.styled';
import { PaginationProps, PaginationTrigger, PaginationType } from './Pagination.types';

const MIN_ALLOWED_VALUE = 1;

export const Pagination = ({
  count,
  currentPage,
  onChange,
  siblingCount,
  type = PaginationType.Default,
  dataTestId,
  disabled = false,
  isWhite = false
}: PaginationProps) => {
  const isInput = type === PaginationType.Input;
  const isCompact = type === PaginationType.Compact;
  const isLastPage = currentPage === count;
  const isFirstPage = currentPage === MIN_ALLOWED_VALUE;
  // TODO: in case this component needs to be used in SSR, include validation for the window or
  // include a custom hook to check for the screen size
  // https://github.com/CarlsbergGBS/cx-component-library/pull/629#discussion_r1187538517
  const buttonSize = window.innerWidth <= 768 ? ButtonSize.Small : ButtonSize.Medium;

  const theme = useContext(ThemeContext) || defaultTheme;

  const [provisionalInputValue, setProvisionalInputValue] = useState<number | string>(currentPage);
  const [inputValue, setInputValue] = useState<number | string>(currentPage);

  const displayedPages = usePagination({
    totalPageCount: count,
    siblingCount,
    currentPage,
    isDefault: type === PaginationType.Default
  });

  useEffect(() => {
    setInputValue(currentPage);
  }, [currentPage]);

  const onPageClick = (targetPage: number) => {
    onChange(targetPage, PaginationTrigger.PageNr);
  };

  const onPageKeyUp =
    (pageNr: number, trigger: PaginationTrigger = PaginationTrigger.PageNr) =>
    (event: KeyboardEvent) => {
      if (event.key === EventKeys.ENTER || event.key === EventKeys.SPACE) {
        event.preventDefault();
        onChange(pageNr, trigger);
      }
    };

  const onPrevious = () => {
    if (type === PaginationType.Input) {
      onChange((inputValue as number) - 1);
      return setInputValue((inputValue as number) - 1);
    }

    return onChange(currentPage - 1, PaginationTrigger.Prev);
  };

  const onNext = () => {
    if (type === PaginationType.Input) {
      onChange((inputValue as number) + 1);
      return setInputValue((inputValue as number) + 1);
    }

    return onChange(currentPage + 1, PaginationTrigger.Next);
  };

  const onPreviousKeyUp = () => {
    onPageKeyUp(currentPage - 1, PaginationTrigger.Prev);
  };

  const onNextKeyUp = () => {
    onPageKeyUp(currentPage + 1, PaginationTrigger.Next);
  };

  const handleOnInputChange = ({ currentTarget: { value } }: ChangeEvent<HTMLInputElement>) => {
    setInputValue(value !== '' ? Number(value) : value);
  };

  const handleAssignInputValue = ({
    currentTarget: { value }
  }: KeyboardEvent<HTMLInputElement> | FocusEvent<HTMLInputElement>) => {
    let newValue = Number(value === '' ? provisionalInputValue : value);

    if (newValue > count) {
      newValue = count;
    } else if (newValue < MIN_ALLOWED_VALUE) {
      newValue = MIN_ALLOWED_VALUE;
    }

    if (newValue !== provisionalInputValue) {
      onChange(newValue, PaginationTrigger.PageNr);
    }

    setInputValue(newValue);
    setProvisionalInputValue(newValue);
  };

  const handleOnKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === EventKeys.ENTER) {
      handleAssignInputValue(event);
    }
  };

  const getTextColor = () => {
    if (isWhite) {
      return TextColor.White;
    }

    if (disabled) {
      return TextColor.DisableLightTheme;
    }

    return TextColor.DigitalBlack;
  };

  const renderContent = () => {
    if (isInput) {
      return (
        <StyledInputPagination theme={theme}>
          <StyledInput
            theme={theme}
            data-testid={`${dataTestId}-input`}
            value={inputValue}
            onChange={handleOnInputChange}
            onBlur={handleAssignInputValue}
            onKeyDown={handleOnKeyDown}
            max={count}
            min={MIN_ALLOWED_VALUE}
            type="number"
            aria-label={`Page ${inputValue}`}
            disabled={disabled}
          />
          <Text
            dataQaId={`${dataTestId}-input-count`}
            textStyle={TextStyle.MediumSmallDefault}
            color={getTextColor()}
            aria-label={`of ${count}`}
          >{` / ${count}`}</Text>
        </StyledInputPagination>
      );
    }

    if (isCompact) {
      return (
        <li>
          <Text
            dataQaId={`${dataTestId}-pagination-compact`}
            textStyle={TextStyle.SmallDefault}
            color={getTextColor()}
            aria-label={`Page ${currentPage} of ${count}`}
          >{`${currentPage} of ${count}`}</Text>
        </li>
      );
    }

    return (
      displayedPages?.map((pageNr) => {
        const isCurrentPage = pageNr === currentPage;
        if (pageNr === LEFT_DOTS || pageNr === RIGHT_DOTS) {
          return (
            <li data-testid={`${dataTestId}-${pageNr}`} key={pageNr} tabIndex={-1}>
              <StyledDots theme={theme} isWhite={isWhite} $disabled={disabled} aria-label="Ellipsis">
                &#8230;
              </StyledDots>
            </li>
          );
        }
        return (
          <li className="default-pagination" key={pageNr}>
            <Button
              dataTestId={`${dataTestId}-page-${pageNr}`}
              style={ButtonStyle.Transparent}
              selected={isCurrentPage}
              onClick={() => onPageClick(Number(pageNr))}
              onKeyUp={() => onPageKeyUp(Number(pageNr))}
              aria-current={isCurrentPage}
              aria-label={isCurrentPage ? `Page ${pageNr}` : `Go to page ${pageNr}`}
              tabIndex={0}
              text={pageNr}
              negative={isWhite}
              size={buttonSize}
              disabled={disabled}
            />
          </li>
        );
      }) ?? null
    );
  };

  return (
    <StyledContainer data-testid={dataTestId} isWhite={isWhite} theme={theme}>
      <ul>
        <li>
          <Button
            dataTestId={`${dataTestId}-button-previous`}
            style={ButtonStyle.Transparent}
            disabled={isFirstPage || disabled}
            tabIndex={isFirstPage ? -1 : 0}
            onClick={onPrevious}
            onKeyUp={onPreviousKeyUp}
            icon={IconName.ChevronLeft}
            size={buttonSize}
            negative={isWhite}
            aria-label="Go to previous page"
          />
        </li>
        {renderContent()}
        <li>
          <Button
            dataTestId={`${dataTestId}-button-next`}
            style={ButtonStyle.Transparent}
            disabled={isLastPage || disabled}
            tabIndex={isLastPage ? -1 : 0}
            onClick={onNext}
            onKeyUp={onNextKeyUp}
            icon={IconName.ChevronRight}
            size={buttonSize}
            negative={isWhite}
            aria-label="Go to next page"
          />
        </li>
      </ul>
    </StyledContainer>
  );
};
