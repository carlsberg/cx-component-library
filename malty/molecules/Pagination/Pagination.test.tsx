import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Pagination } from './Pagination';
import { PaginationType } from './Pagination.types';

describe('pagination', () => {
  it('renders elements', () => {
    render(<Pagination currentPage={1} count={10} dataQaId="pagination" onChange={() => null} />);
    expect(screen.getByTestId('pagination')).toBeInTheDocument();
  });
  it('renders current page', () => {
    const currentPage = 3;
    render(
      <Pagination
        currentPage={currentPage}
        count={10}
        dataQaId="pagination"
        onChange={() => null}
      />
    );
    expect(screen.getByTestId('pagination')).toBeInTheDocument();
    const el = screen.getByText(currentPage);
    expect(el.innerHTML).toEqual(currentPage.toString());
  });
  it('renders input type', () => {
    const currentPage = 3;
    render(
      <Pagination
        currentPage={currentPage}
        count={10}
        type={PaginationType.Input}
        dataQaId="pagination"
        onChange={() => null}
      />
    );
    expect(screen.getByTestId('pagination-input')).toBeInTheDocument();
    const el = screen.getByDisplayValue(currentPage);
    expect(el.getAttribute('value')).toEqual(currentPage.toString());
  });
  it('calls onChange function', () => {
    const onChange = jest.fn();
    const currentPage = 1;
    render(
      <Pagination currentPage={currentPage} count={10} dataQaId="pagination" onChange={onChange} />
    );
    fireEvent.click(screen.getByTestId('pagination-button-next'));
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
