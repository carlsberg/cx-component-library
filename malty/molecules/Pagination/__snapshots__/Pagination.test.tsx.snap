// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecule pagination matches snapshot 1`] = `
<div
  className="sc-kDDrLX kGuhpW"
>
  <span
    className="sc-bczRLJ hROJNc"
  >
    <ul>
      <li>
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ hQuGSn cUKKxM"
            data-testid="undefined-button-previous"
            disabled={true}
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={-1}
            type="button"
          >
            <div
              className="text-container "
            >
              <svg
                className="sc-gsnTZi gIwTyE"
                color="#212833"
                data-testid="svg-component"
                size="24px"
                version="1.1"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
                xmlnsXlink="http://www.w3.org/1999/xlink"
              >
                <g
                  fillRule="evenodd"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M10.414 12l5.293-5.293a1 1 0 00-1.414-1.414l-6 6a1 1 0 000 1.414l6 6a1 1 0 101.414-1.414z"
                  />
                </g>
              </svg>
            </div>
          </button>
        </span>
      </li>
      <li
        className="default-pagination"
      >
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ foJVnI cUKKxM active"
            data-testid="undefined-page-1"
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <div
              className="text-container "
            >
              1
            </div>
          </button>
        </span>
      </li>
      <li
        className="default-pagination"
      >
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ foJVnI cUKKxM"
            data-testid="undefined-page-2"
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <div
              className="text-container "
            >
              2
            </div>
          </button>
        </span>
      </li>
      <li
        className="default-pagination"
      >
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ foJVnI cUKKxM"
            data-testid="undefined-page-3"
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <div
              className="text-container "
            >
              3
            </div>
          </button>
        </span>
      </li>
      <li
        className="default-pagination"
      >
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ foJVnI cUKKxM"
            data-testid="undefined-page-4"
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <div
              className="text-container "
            >
              4
            </div>
          </button>
        </span>
      </li>
      <li
        className="default-pagination"
      >
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ foJVnI cUKKxM"
            data-testid="undefined-page-5"
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <div
              className="text-container "
            >
              5
            </div>
          </button>
        </span>
      </li>
      <li
        data-testid="undefined-dots"
        tabIndex={-1}
      >
        <div
          className="sc-iqcoie lcUdKR"
        >
          â€¦
        </div>
      </li>
      <li
        className="default-pagination"
      >
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ foJVnI cUKKxM"
            data-testid="undefined-page-10"
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <div
              className="text-container "
            >
              10
            </div>
          </button>
        </span>
      </li>
      <li>
        <span
          className="sc-bczRLJ hROJNc"
        >
          <button
            className="sc-gKXOVf sc-papXJ hQuGSn cUKKxM"
            data-testid="undefined-button-next"
            disabled={false}
            fontSize="16px"
            loading={false}
            onClick={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <div
              className="text-container "
            >
              <svg
                className="sc-gsnTZi gIwTyE"
                color="#212833"
                data-testid="svg-component"
                size="24px"
                version="1.1"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
                xmlnsXlink="http://www.w3.org/1999/xlink"
              >
                <g
                  fillRule="evenodd"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M13.586 12l-5.293 5.293a1 1 0 001.414 1.414L16.414 12 9.707 5.293a1 1 0 00-1.414 1.414z"
                  />
                </g>
              </svg>
            </div>
          </button>
        </span>
      </li>
    </ul>
  </span>
</div>
`;
